### PR TITLE
bpo-40222: "Zero cost" exception handling

### PR DESCRIPTION
**Pull request by markshannon**
For example, this function:
```
def f():
    try:
        g(0)
    except:
        return "fail"
```
compiles as follows on main:
```
  2           0 SETUP_FINALLY            7 (to 16)

  3           2 LOAD_GLOBAL              0 (g)
              4 LOAD_CONST               1 (0)
              6 CALL_FUNCTION            1
              8 POP_TOP
             10 POP_BLOCK
             12 LOAD_CONST               0 (None)
             14 RETURN_VALUE

  4     >>   16 POP_TOP
             18 POP_TOP
             20 POP_TOP

  5          22 POP_EXCEPT
             24 LOAD_CONST               3 ('fail')
             26 RETURN_VALUE
```

with this PR it compiles as follows:
```
  2           0 NOP

  3           2 LOAD_GLOBAL              0 (g)
              4 LOAD_CONST               1 (0)
              6 CALL_FUNCTION            1
              8 POP_TOP
             10 LOAD_CONST               0 (None)
             12 RETURN_VALUE
        >>   14 PUSH_EXC_INFO

  4          16 POP_TOP
             18 POP_TOP
             20 POP_TOP

  5          22 POP_EXCEPT
             24 LOAD_CONST               2 ('fail')
             26 RETURN_VALUE
        >>   28 POP_EXCEPT_AND_RERAISE
ExceptionTable:
  2 to 8 -> 14 [0]
  14 to 20 -> 28 [3] lasti
```

This is not quite zero-cost at the moment, as it leaves a `NOP`s for each `try`, and possibly a few other.

Removing `NOP`s that are on a line by themselves will require changes to the line number table, which has nothing to do with exception handling and I don't want to mix the two PRs.

Although the code is slightly longer overall, there is less work done on the non-exceptional path, one `NOP` versus a `SETUP_FINALLY` and `POP_BLOCK`.

Not only is there a reduction in work done in the bytecode, but in the size of frames is reduced a lot:
```
def frame_size():
    return sys.getsizeof(sys._getframe())
```
On master:
```
>>> frame_size()
400
```
with this PR:
```
>>> frame_size()
152
```

https://bugs.python.org/issue40222

### CODE REVIEW

File: Include/cpython/frameobject.h [hunk 1]
```diff
@@ -34,18 +34,14 @@ struct _frame {
...
-
     /* Borrowed reference to a generator, or NULL */
     PyObject *f_gen;
-
+    int f_stackdepth;           /* Depth of value stack */
     int f_lasti;                /* Last instruction if called */
     int f_lineno;               /* Current line number. Only valid if non-zero */
-    int f_iblock;               /* index in f_blockstack */
     PyFrameState f_state;       /* What state the frame is in */
-    PyTryBlock f_blockstack[CO_MAXBLOCKS]; /* for try and loop blocks */
```
Comment (ericsnowcurrently) [inline Include/cpython/frameobject.h:48]: So we already weren't using this for loop blocks?

File: Lib/importlib/_bootstrap_external.py [hunk 2]
```diff
@@ -361,7 +362,7 @@ def _write_atomic(path, data, mode=0o666):
 # Whenever MAGIC_NUMBER is changed, the ranges in the magic_values array
 # in PC/launcher.c must also be updated.
 
-MAGIC_NUMBER = (3439).to_bytes(2, 'little') + b'\r\n'
+MAGIC_NUMBER = (3450).to_bytes(2, 'little') + b'\r\n'
```
Comment (ericsnowcurrently) [inline Lib/importlib/_bootstrap_external.py:365]: Did you mean to use `3440` instead?

File: Misc/NEWS.d/next/Core and Builtins/2021-04-30-15-48-36.bpo-40222.j3VxeX.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+"Zero cost" exception handling.
+
+* Uses a lookup table to determine how to handle exceptions
+* Removes SETUP_FINALLY and POP_TOP block instructions, eliminating the runtime overhead of try statements.
+* Reduces the size of the fram object by about 60%
```
Comment (ericsnowcurrently) [inline Misc/NEWS.d/next/Core and Builtins/2021-04-30-15-48-36.bpo-40222.j3VxeX.rst]:
```suggestion
* Uses a lookup table to determine how to handle exceptions.
* Removes SETUP_FINALLY and POP_TOP block instructions, eliminating the runtime overhead of try statements.
* Reduces the size of the frame object by about 60%.
```

or 

```suggestion
* uses a lookup table to determine how to handle exceptions
* removes SETUP_FINALLY and POP_TOP block instructions, eliminating the runtime overhead of try statements
* reduces the size of the frame object by about 60%
```

Comment (ericsnowcurrently) [inline Misc/NEWS.d/next/Core and Builtins/2021-04-30-15-48-36.bpo-40222.j3VxeX.rst]:
```suggestion
* Reduces the size of the frame object by about 60%
```

File: Include/cpython/frameobject.h [hunk 4]
```diff
@@ -77,11 +73,6 @@ _PyFrame_New_NoTrack(PyThreadState *, PyFrameConstructor *, PyObject *);
 
 /* The rest of the interface is specific for frame objects */
 
-/* Block management functions */
-
-PyAPI_FUNC(void) PyFrame_BlockSetup(PyFrameObject *, int, int, int);
-PyAPI_FUNC(PyTryBlock *) PyFrame_BlockPop(PyFrameObject *);
```
Comment (ericsnowcurrently) [inline Include/cpython/frameobject.h:83]: Given that these were part of the public C-API, it would probably be worth adding a note to the What's New doc about their removal.  I suppose the same may be true of `f_blockstack`, though it's less likely that matters.

File: Tools/scripts/generate_opcode_h.py [hunk 5]
```diff
@@ -16,14 +16,10 @@
...
 footer = """
-/* EXCEPT_HANDLER is a special, implicit block type which is created when
-   entering an except handler. It is not an opcode but we define it here
-   as we want it to be available to both frameobject.c and ceval.c, while
-   remaining private.*/
-#define EXCEPT_HANDLER 257
-
 #define HAS_ARG(op) ((op) >= HAVE_ARGUMENT)
 
+#define IS_ARTIFICIAL(op) ((op) > 240)
```
Comment (ericsnowcurrently) [inline Tools/scripts/generate_opcode_h.py:23]: What is special about 240?  A note here would go along way toward making this line less mysterious.  It may be more useful to put such a note (and the specific value) in opcode.py instead, since the relationship with the data there is much more obvious.

File: Objects/exception_table_notes.txt [hunk 6]
```diff
@@ -0,0 +1,82 @@
+Description of the internal format of the exception table.
+
+Conceptually, the exception table consists of a sequence of 5-tuples:
+    start-offset (inclusive), end-offset (exclusive), target, stack-depth, push-lasti (boolean)
```
Comment (ericsnowcurrently) [inline Objects/exception_table_notes.txt]: For a little more readability:

```suggestion
Conceptually, the exception table consists of a sequence of 5-tuples:
    1. start-offset (inclusive)
    2. end-offset (exclusive)
    3. target
    4. stack-depth
    5. push-lasti (boolean)
```

File: Objects/exception_table_notes.txt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    depth  3
+    lasti  False
+
+is encoded first by converting to the more compact four value form:
+    start:        20
+    size:         8
+    target        100
+  depth<<1+lasti  6
+
+Where is then encoded as:
```
Comment (ericsnowcurrently) [inline Objects/exception_table_notes.txt]:
```suggestion
Which is then encoded as:
```

File: Objects/exception_table_notes.txt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            start = parse_varint(iterator)*2
+            length = parse_varint(iterator)*2
+            end = start + length - 2 # Present as inclusive, not exclusive
+            target = parse_varint(iterator)*2
+            dl = parse_varint(iterator)
+            depth = dl >> 1
+            lasti = bool(dl&1)
+            yield start, end, target, depth, lasti
+    except StopIteration:
+        return
```
Comment (ericsnowcurrently) [inline Objects/exception_table_notes.txt]: Could you just point to dis.py instead?

File: Objects/exception_table_notes.txt [hunk 9]
```diff
@@ -0,0 +1,82 @@
+Description of the internal format of the exception table.
```
Comment (ericsnowcurrently) [inline Objects/exception_table_notes.txt]: FWIW, this file is really helpful.

File: Objects/frameobject.c [hunk 10]
```diff
@@ -132,15 +137,16 @@ markblocks(PyCodeObject *code_obj, int len)
         PyErr_NoMemory();
         return NULL;
     }
-    memset(blocks, -1, (len+1)*sizeof(int64_t));
+    for (int i = 1; i <= len; i++) {
+        blocks[i] = UNINITIALIZED;
+    }
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c]: That's definitely more comprehensible. 🙂

File: Objects/frameobject.c [hunk 11]
```diff
@@ -92,26 +92,31 @@ get_arg(const _Py_CODEUNIT *codestr, Py_ssize_t i)
...
-    Except = 4,
+    Iterator = 1,
+    Except = 2,
+    Object = 3,
 } Kind;
 
-#define BITS_PER_BLOCK 3
+#define BITS_PER_BLOCK 2
+
+#define UNINITIALIZED -2
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c:108]: Yeah, adding this definitely helps make the rest of the code more readable.

File: Objects/frameobject.c [hunk 12]
```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
...
+                {
+                    int j = get_arg(code, i);
                     assert(j < len);
-                    if (blocks[j] == -1 && j < i) {
+                    if (blocks[j] == UNINITIALIZED && j < i) {
                         todo = 1;
                     }
-                    assert(blocks[j] == -1 || blocks[j] == block_stack);
-                    blocks[j] = block_stack;
+                    if (opcode == JUMP_IF_NOT_EXC_MATCH) {
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c:175]: Ah, right, we fell through so this might be false.

File: Objects/frameobject.c [hunk 13]
```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
...
+                        block_stack = pop_block(block_stack);
+                    }
+                    int64_t target_stack = block_stack;
+                    if (opcode != JUMP_IF_FALSE_OR_POP && opcode != JUMP_IF_TRUE_OR_POP) {
+                        target_stack = pop_block(target_stack);
+                    }
+                    block_stack = pop_block(block_stack);
+                    assert(blocks[j] == UNINITIALIZED || blocks[j] == target_stack);
+                    blocks[j] = target_stack;
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c]: FWIW, the split here with `target_stack` is a little hard to follow at first.  A slight change would help with that:

```suggestion
                    int64_t target_stack = block_stack;
                    if (opcode == JUMP_IF_NOT_EXC_MATCH) {
                        block_stack = pop_block(block_stack);
                        target_stack = block_stack;
                    }
                    if (opcode != JUMP_IF_FALSE_OR_POP && opcode != JUMP_IF_TRUE_OR_POP) {
                        target_stack = pop_block(target_stack);
                    }
                    block_stack = pop_block(block_stack);
                    assert(blocks[j] == UNINITIALIZED || blocks[j] == target_stack);
                    blocks[j] = target_stack;
```

...though it still feels off somehow.  Perhaps a comment would be worth adding here to make sure it is clear what is happening?

File: Objects/frameobject.c [hunk 14]
```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
...
-                case SETUP_WITH:
-                case SETUP_ASYNC_WITH:
-                    j = get_arg(code, i) + i + 1;
-                    assert(j < len);
-                    except_stack = push_block(block_stack, Except);
-                    assert(blocks[j] == -1 || blocks[j] == except_stack);
-                    blocks[j] = except_stack;
-                    block_stack = push_block(block_stack, With);
+                case POP_EXCEPT:
+                    block_stack = pop_block(pop_block(pop_block(block_stack)));
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c]: Why do we pop 3?

File: Objects/frameobject.c [hunk 15]
```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
...
                     assert(j < len);
-                    assert(blocks[j] == -1 || blocks[j] == block_stack);
-                    blocks[j] = block_stack;
-                    break;
-                case POP_BLOCK:
-                case POP_EXCEPT:
-                    block_stack = pop_block(block_stack);
-                    blocks[i+1] = block_stack;
+                    assert(blocks[j] == UNINITIALIZED || blocks[j] == target_stack);
+                    blocks[j] = target_stack;
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c]: Again, it might be helpful to have a comment that makes the division between `block_stack` and `target_stack` more clear.

File: Objects/frameobject.c [hunk 16]
```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
...
+                    blocks[i+1] = block_stack;
                 case RETURN_VALUE:
                 case RAISE_VARARGS:
                 case RERAISE:
+                case POP_EXCEPT_AND_RERAISE:
                     /* End of block */
                     break;
+                case GEN_START:
+                    blocks[i+1] = block_stack;
+                    break;
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c]: Why is this necessary, where before it wasn't?

File: Objects/frameobject.c [hunk 17]
```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
...
+                {
+                    int delta = PyCompile_OpcodeStackEffect(opcode, _Py_OPARG(code[i]));
+                    while (delta < 0) {
+                        block_stack = pop_block(block_stack);
+                        delta++;
+                    }
+                    while (delta > 0) {
+                        block_stack = push_block(block_stack, Object);
+                        delta--;
+                    }
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c]: These loops are mutually exclusive, right?

Comment (ericsnowcurrently) [inline Objects/frameobject.c]: Why was this not necessary before?

File: Objects/frameobject.c [hunk 18]
```diff
@@ -150,109 +156,144 @@ markblocks(PyCodeObject *code_obj, int len)
...
 
 static const char *
 explain_incompatible_block_stack(int64_t to_stack)
 {
+    assert(to_stack != 0);
+    if (to_stack == OVERFLOWED) {
+        return "stack is too deep to analyze";
+    }
+    if (to_stack == UNINITIALIZED) {
+        return "can't jump into an exception handler, or code may be unreachable.";
```
Comment (ericsnowcurrently) [inline Objects/frameobject.c]:
```suggestion
        return "can't jump into an exception handler, or code may be unreachable";
```

File: Python/ceval.c [hunk 19]
```diff
@@ -1800,14 +1770,14 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
...
-            stack_pointer = f->f_valuestack+f->f_stackdepth;
-            f->f_stackdepth = -1;
             if (err) {
                 /* trace function raised an exception */
                 goto error;
             }
+            /* Reload possibly changed frame fields */
+            JUMPTO(f->f_lasti);
+            stack_pointer = f->f_valuestack+f->f_stackdepth;
+            f->f_stackdepth = -1;
```
Comment (ericsnowcurrently) [inline Python/ceval.c:1780]: Did you move this as a consequence of this PR or is it something you happened to notice?

File: Python/ceval.c [hunk 20]
```diff
@@ -4081,22 +4057,19 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
...
         case TARGET(WITH_EXCEPT_START): {
-            /* At the top of the stack are 7 values:
+            /* At the top of the stack are 8 values:
                - (TOP, SECOND, THIRD) = exc_info()
-               - (FOURTH, FIFTH, SIXTH) = previous exception for EXCEPT_HANDLER
-               - SEVENTH: the context.__exit__ bound method
+               - (FOURTH, FIFTH, SIXTH) = previous exception
+               - SEVENTH: lasti of exception in exc_info()
+               - EIGHTH: the context.__exit__ bound method
                We call SEVENTH(TOP, SECOND, THIRD).
```
Comment (ericsnowcurrently) [inline Python/ceval.c]:
```suggestion
               We call EIGHTH(TOP, SECOND, THIRD).
```

File: Python/ceval.c [hunk 21]
```diff
@@ -4777,6 +4772,96 @@ positional_only_passed_as_keyword(PyThreadState *tstate, PyCodeObject *co,
 
 }
 
+static inline unsigned char *
+parse_varint(unsigned char *p, int *result) {
+    int val = p[0]&63;
```
Comment (ericsnowcurrently) [inline Python/ceval.c]:
```suggestion
    int val = p[0] & 63;
```

File: Python/ceval.c [hunk 22]
```diff
@@ -4777,6 +4772,96 @@ positional_only_passed_as_keyword(PyThreadState *tstate, PyCodeObject *co,
 
 }
 
+static inline unsigned char *
```
Comment (ericsnowcurrently) [inline Python/ceval.c:4779]: It would be worth adding a comment here that points back to exception_table_notes.txt.

File: Python/compile.c [hunk 23]
```diff
@@ -67,17 +67,38 @@
  */
 #define MAX_ALLOWED_STACK_USE (STACK_USE_GUIDELINE * 100)
 
+
+/* Pseudo-instructions used in the compiler,
+ * but turned into NOPs by the assembler. */
+#define SETUP_FINALLY 255
+#define SETUP_CLEANUP 254
+#define SETUP_WITH 253
+#define POP_BLOCK 252
```
Comment (ericsnowcurrently) [inline Python/compile.c:76]: Is this a temporary situation?

File: Python/compile.c [hunk 24]
```diff
@@ -101,7 +122,8 @@ is_relative_jump(struct instr *i)
 static inline int
 is_jump(struct instr *i)
 {
-    return is_bit_set_in_table(_PyOpcode_Jump, i->i_opcode);
+
```
Comment (ericsnowcurrently) [inline Python/compile.c]:
```suggestion
```

File: Python/compile.c [hunk 25]
```diff
@@ -1116,9 +1143,14 @@ stack_effect(int opcode, int oparg, int jump)
             /* 0 in the normal flow.
              * Restore the stack position and push 6 values before jumping to
```
Comment (ericsnowcurrently) [inline Python/compile.c]:
```suggestion
             * Restore the stack position and push 3 values before jumping to
```

File: Python/compile.c [hunk 26]
```diff
@@ -1116,9 +1143,14 @@ stack_effect(int opcode, int oparg, int jump)
...
              * the handler if an exception be raised. */
-            return jump ? 6 : 0;
+            return jump ? 3 : 0;
+        case SETUP_CLEANUP:
+            return jump ? 4 : 0;
```
Comment (ericsnowcurrently) [inline Python/compile.c:1148]: I'm guessing 3 of those are the same as SETUP_FINALLY.  What's the fourth?

File: Python/compile.c [hunk 27]
```diff
@@ -1165,13 +1197,15 @@ stack_effect(int opcode, int oparg, int jump)
         /* Iterators and generators */
         case GET_AWAITABLE:
             return 0;
-        case SETUP_ASYNC_WITH:
+        case SETUP_WITH:
             /* 0 in the normal flow.
              * Restore the stack position to the position before the result
-             * of __aenter__ and push 6 values before jumping to the handler
+             * of __aenter__ and push 7 values before jumping to the handler
```
Comment (ericsnowcurrently) [inline Python/compile.c]:
```suggestion
             * of __aenter__ and push 4 values before jumping to the handler
```

Comment (ericsnowcurrently) [reply Python/compile.c]: Also, is that supposed to be `__enter__` instead of `__aenter__`?

File: Python/compile.c [hunk 28]
```diff
@@ -3159,15 +3203,20 @@ compiler_try_except(struct compiler *c, stmt_ty s)
...
         return 0;
     for (i = 0; i < n; i++) {
         excepthandler_ty handler = (excepthandler_ty)asdl_seq_GET(
             s->v.Try.handlers, i);
-        if (!handler->v.ExceptHandler.type && i < n-1)
-            return compiler_error(c, "default 'except:' must be last");
         SET_LOC(c, handler);
+        if (!handler->v.ExceptHandler.type && i < n-1) {
+            return compiler_error(c, "default 'except:' must be last");
+        }
```
Comment (ericsnowcurrently) [inline Python/compile.c:3219]: Is this another case where you happened to notice it was incorrect, or is this change due to this PR?

File: Python/compile.c [hunk 29]
```diff
@@ -4764,6 +4818,8 @@ compiler_async_comprehension_generator(struct compiler *c,
     compiler_pop_fblock(c, ASYNC_COMPREHENSION_GENERATOR, start);
 
     compiler_use_next_block(c, except);
+    c->u->u_lineno = -1;
```
Comment (ericsnowcurrently) [inline Python/compile.c]: Likewise, is this due to the other changes or something you happened to notice?

File: Python/compile.c [hunk 30]
```diff
@@ -6383,11 +6454,13 @@ compiler_match(struct compiler *c, stmt_ty s)
 */
 
 struct assembler {
-    PyObject *a_bytecode;  /* string containing bytecode */
+    PyObject *a_bytecode;  /* bytes containing bytecode */
     int a_offset;              /* offset into bytecode */
     int a_nblocks;             /* number of reachable blocks */
-    PyObject *a_lnotab;    /* string containing lnotab */
+    PyObject *a_lnotab;    /* bytes containing lnotab */
```
Comment (ericsnowcurrently) [inline Python/compile.c:6460]: 😄

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Mostly LGTM, but with a bunch of small comments to sort out first

My only big request would be a bit more explanation somewhere of how the zero-cost exception handling works.  I didn't have much context when reading through frameobject.c, as you may note by a number of "why did this change" comments. 🙂

Also, have you been able to benchmark this?  My guess is that it will be faster and leaner generally, but a little slower when handling exceptions.

File: Include/cpython/frameobject.h (hunk 1 above)
Comment (markshannon) [reply Include/cpython/frameobject.h:48]: Haven't used this for for loops since 3.8.
https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Doc/whatsnew/3.8.rst#cpython-bytecode-changes

File: Lib/importlib/_bootstrap_external.py (hunk 2 above)
Comment (markshannon) [reply Lib/importlib/_bootstrap_external.py:365]: No, we need to leave space for 3.10 fixes.

File: Objects/exception_table_notes.txt (hunk 8 above)
Comment (markshannon) [reply Objects/exception_table_notes.txt]: The code in dis has a slightly different purpose.
This code is (at least should be) as clear as possible, the code in `dis` might have performance or other considerations.
Also, I'd like to keep this self-contained.

File: Objects/frameobject.c (hunk 14 above)
Comment (markshannon) [reply Objects/frameobject.c]: Because that's what POP_EXCEPT does; it pops three values from the *evaluation* stack.
I suspect that would have clearer if I'd renamed `block_stack` :slightly_smiling_face:

File: Objects/frameobject.c (hunk 16 above)
Comment (markshannon) [reply Objects/frameobject.c]: Because we used to model the block stack, now we model the evaluation stack.

File: Objects/frameobject.c (hunk 17 above)
Comment (markshannon) [reply Objects/frameobject.c]: yes

Comment (markshannon) [reply Objects/frameobject.c]: Because we used to model the block stack, now we model the evaluation stack.

File: Python/ceval.c (hunk 19 above)
Comment (markshannon) [reply Python/ceval.c:1780]: As a consequence of this PR. 
Before this PR, unwinding was determined by the block stack, so it didn't matter if we did the `JUMPTO(f->f_lasti)` before unwinding.
However, now we rely on `next_instr` to look up the exception handler, we can't change `next_instr`.

File: Python/compile.c (hunk 23 above)
Comment (markshannon) [reply Python/compile.c:76]: Maybe. I don't plan on changing it, but others might.
These are internal to the compiler. The rest of the code base shouldn't need to to care how we get from AST+symbols to bytecode.

File: Python/compile.c (hunk 26 above)
Comment (markshannon) [reply Python/compile.c:1148]: `lasti`, the instruction offset where the exception was raised.

File: Python/compile.c (hunk 27 above)
Comment (markshannon) [reply Python/compile.c]: Either, as it is preceded by either `BEFORE_WITH` or `BEFORE_ASYNC_WITH`.

File: Python/compile.c (hunk 28 above)
Comment (markshannon) [reply Python/compile.c:3219]: I think this was incorrect before. I'll see if it makes sense to apply it to 3.10b.

File: Python/compile.c (hunk 29 above)
Comment (markshannon) [reply Python/compile.c]: I've reverted this change. Handling of re-raising exceptions needs to be fixed for both 3.10 and 3.11, but will need different treatments. See https://bugs.python.org/issue43933

File: Objects/frameobject.c (hunk 14 above)
Comment (ericsnowcurrently) [reply Objects/frameobject.c]: Yeah, changing the names from "block" to "stack" helps a lot.  Thanks!

File: Objects/frameobject.c (hunk 13 above)
Comment (ericsnowcurrently) [reply Objects/frameobject.c]: Your updates make this section perfectly clear now.  Thanks!

**Comment by markshannon** [comment]
The failure on `buildbot/AMD64 Arch Linux Asan PR` is a pre-existing failure.
The failure on `buildbot/PPC64LE Fedora Stable` seems to be some sort of race condition in `concurrent_futures`

**Comment by pablogsal** [comment]
> The failure on `buildbot/AMD64 Arch Linux Asan PR` is a pre-existing failure.
> The failure on `buildbot/PPC64LE Fedora Stable` seems to be some sort of race condition in `concurrent_futures`

@markshannon Unfortunately the ASAN failure is legit. Check the BPO issue for more details.

File: Include/cpython/code.h [hunk 31]
```diff
@@ -117,12 +118,12 @@ PyAPI_DATA(PyTypeObject) PyCode_Type;
...
         int, int, int, int, int, PyObject *, PyObject *,
         PyObject *, PyObject *, PyObject *, PyObject *,
-        PyObject *, PyObject *, int, PyObject *);
+        PyObject *, PyObject *, int, PyObject *, PyObject *);
 
 PyAPI_FUNC(PyCodeObject *) PyCode_NewWithPosOnlyArgs(
         int, int, int, int, int, int, PyObject *, PyObject *,
         PyObject *, PyObject *, PyObject *, PyObject *,
-        PyObject *, PyObject *, int, PyObject *);
+        PyObject *, PyObject *, int, PyObject *, PyObject *);
```
Comment (scoder) [inline Include/cpython/code.h:126]: This C-API change seems worth mentioning somewhere.

Comment (pablogsal) [reply Include/cpython/code.h:126]: Check the discussion in https://bugs.python.org/issue40222

File: Lib/dis.py [hunk 32]
```diff
@@ -308,8 +311,33 @@ def _get_name_info(name_index, name_list):
     return argval, argrepr
 
 
+def parse_varint(iterator):
+    b = next(iterator)
+    val = b & 63
```
Comment (Kenan7) [inline Lib/dis.py:316]: Just curious, how does this work? I guess the constant numbers are about 64-bit? 

would love to get your explanation, thanks. @markshannon

Comment (ColsonXu-Hedgineer) [reply Lib/dis.py:316]: .
